client-sdk/ts-web/core/src/hdkey.ts

```diff
@@ -0,0 +1,107 @@
+import {sha512} from 'js-sha512';
+import {SignKeyPair, sign} from 'tweetnacl';
+import {generateMnemonic, mnemonicToSeed} from 'bip39';
+import {concat} from './misc';
+
+const ED25519_CURVE = 'ed25519 seed';
+const HARDENED_OFFSET = 0x80000000;
+const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
+
+export class HDKey {
+    public readonly keypair: SignKeyPair;
+
+    constructor(public readonly privateKey: Uint8Array, public readonly chainCode: Uint8Array) {
+        this.keypair = sign.keyPair.fromSeed(privateKey);
+    }
+
+    /**
+     * Generates the keypair matching the supplied parameters
+     * @param mnemonic Bip39 Mnemonic
+     * @param index Account index
+     * @param passphrase Optional Bip39 passphrase
+     * @returns SignKeyPair for these parameters
+     */
+    public static async getAccountSigner(
+        mnemonic: string,
+        index: number = 0,
+        passphrase?: string,
+    ): Promise<SignKeyPair> {
+        const key = await HDKey.fromMnemonic(mnemonic, passphrase);
+        return key.derivePath(`m/44'/474'/${index}'`).keypair;
+    }
+
+    /**
+     * Returns an instance of HDKey that can then be used for derivation
+     * @param seed Master seed for key derivation
+     * @returns Instance of HDKey
+     */
+    public static fromSeed(seed: Uint8Array): HDKey {
+        return HDKey.makeHDKey(ED25519_CURVE, seed);
+    }
+
+    /**
+     * Returns the root node for the given mnemonic
+     * @param mnemonic Bip39 Mnemonic
+     * @param passphrase Optional passphrase
+     * @returns Root node
+     */
+    public static async fromMnemonic(mnemonic: string, passphrase?: string): Promise<HDKey> {
+        const seed = await mnemonicToSeed(mnemonic, passphrase);
+        return HDKey.fromSeed(seed);
+    }
+
+    /**
+     * Generates a mnemonic
+     * @param strength Length in bits of the generated mnemonic
+     * @returns Generated BIP39 Mnemonic
+     */
+    public static generateMnemonic(strength = 256): string {
+        return generateMnemonic(strength);
+    }
+
+    /**
+     * Returns the HDKey for the given derivation path
```

Comment:
I would mention SLIP-0010 in this docstring and the docstrings of other functions that implement SLIP-0010.